cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1803,37 +1805,63 @@ class AzureFileSystem::Impl {
                            const AzureLocation& location, bool recursive) {
     DCHECK(!location.container.empty());
     DCHECK(!location.path.empty());
-    // Non-recursive CreateDir calls require the parent directory to exist.
-    if (!recursive) {
+    if (recursive) {
+      std::vector<AzureLocation> target_locations;
+      // Recursive CreateDir calls require there is no file in parents.
+      auto parent = location;
+      while (!parent.path.empty()) {
+        ARROW_ASSIGN_OR_RAISE(auto info, GetFileInfo(container_client, parent));
+        if (info.type() == FileType::File) {
+          return NotADir(parent);
+        }
+        if (info.type() == FileType::NotFound) {
+          target_locations.push_back(parent);
+        }
+        parent = parent.parent();
+        if (parent.path.empty() && info.type() == FileType::NotFound) {
+          // parent is container and may not exist
+          ARROW_ASSIGN_OR_RAISE(info,
+                                GetContainerPropsAsFileInfo(parent, container_client));
+          if (info.type() == FileType::NotFound) {
+            try {
+              container_client.CreateIfNotExists();
+            } catch (const Storage::StorageException& exception) {
+              return ExceptionToStatus(exception, "Failed to create directory '",
+                                       location.all, "': ", container_client.GetUrl());
+            }
+          }
+        }
+      }
+      for (size_t i = target_locations.size(); i > 0; --i) {
+        const auto& target_location = target_locations[i - 1];
+        try {
+          create_if_not_exists(container_client, target_location);
+        } catch (const Storage::StorageException& exception) {
+          return ExceptionToStatus(exception, "Failed to create directory '",
+                                   location.all, "': ", container_client.GetUrl());
```

Comment:
After the first iteration of this loop, an exception shouldn't report a failure of the entire operation: the full path creates all the prefixes (they are "implied directories"). All the other functions handle implied directories, so it should be OK to rely on them.